### DISCUSSION TITLE
using databridge - failing to send messages to a REST API

### DISCUSSION THREAD

**Discussion by vibhubithar** [emqx/emqx#9808] [Q&A] [answered]
Hi all,

i am using the open-source version where i am using a data bridge to receive messages from an external MQTT broker (which is working fine) but when it comes to sending the payload out to a REST API i am not getting any data.

When looking at the GUI the ingress and egress are confusing me. Do I need to enable the egress topic so as to enable the webhook ? or I can take the payload from the external broker and forward it to the REST API

**Comment by lafirest** [comment]
@vibhubithar 
I think you can have a try with the newest version, v5.0.14, it simplifies the bridges config.
And if you got something except than what you want, you can set the log level to `debug`, this could help you to find what happened.

**Comment by vibhubithar** [comment]
thanks, i am using "emqx/emqx:5.0.14"

What i am not sure is, whether can i take the payload from an external broker and send it to API directly via webhook

external messagebroker -> EMQX data bridge -> SQL -> Webhook -> external API

OR

1st - external messagebroker -> EMQX data bridge -> SQL -> local broker 
then - local broker ->  Webhook -> external API

i am following the example you have here

 https://www.emqx.io/docs/en/v5.0/data-integration/data-bridges.html#quick-start-with-data-bridge

**Comment by vibhubithar** [comment]
so i am able to get messages from external broker to EMQX data bridge and even see the message in local topic

But i cant get the payload be sent to API i created based on the example provided

![simple-server-webhook](https://user-images.githubusercontent.com/10491706/213350756-5b50d01d-2c99-46f4-b975-86f56f3608d3.png)


![flow with local broker to simpleserver](https://user-images.githubusercontent.com/10491706/213350798-6575861a-e06a-49e0-87a0-aae5b3ecb164.png)

![databridge-webhook-no data to api](https://user-images.githubusercontent.com/10491706/213350896-b2dcae9d-bae8-4e9c-a99e-4957b6eb5df3.png)



![rule-databridge-sql select](https://user-images.githubusercontent.com/10491706/213350925-d06ad7a2-c7de-43f5-8591-9e537d16b0c2.png)

**Reply by lafirest** [reply]
From the first picture, we can see there may be some network problems between the EMQX and your web server.
EMQX can't connect to your server and seems all data were queued.

**Comment by vibhubithar** [comment, marked as answer]
the webserver is working fine, I posted a message to it using the POSTMAN application.

I don't have to use an API key or anything, so looks like for some reason EMQX webhook is facing some issues, could totally be my config but o can't point out what's missing in EMQX webhook settings

![postman-webserver-api-working](https://user-images.githubusercontent.com/10491706/213374579-7937d4d7-6284-4b78-97cf-eb37a5995d82.png)

**Reply by lafirest** [reply]
Please set the EMQX log level to `debug`, it will print the reason why the connect failed.

**Reply by vibhubithar** [reply]
so i used a dummy API server address on internet and EMQX was able to post to that API

Could it be an issue with the way i am running EMQX in docker?

As a result when i try to set API Address to localhost in EMQX it tries to find an API server inside the docker which doesn't exist

**Reply by vibhubithar** [reply]
then the question becomes how the example tested it, was it not using docker to run EMQX?

**Reply by vibhubithar** [reply]
So the problem is the example assumes you are running EMQX natively on the machine as a result the simple webserver python example launches the API server on localhost

Here is my sample API server code 

from flask import Flask, json, request
import os

api = Flask(__name__)

@api.route('/', methods=['POST'])
def print_messages():
  reply= {"result": "ok", "message": "success"}
  print("got post request: ", request.get_data())
  return json.dumps(reply), 200

if __name__ == '__main__':
  port = int(os.environ.get("PORT", 5000))
  api.run(host='10.0.0.126', port=port)
  # api.run(debug=True)

**Reply by lafirest** [reply]
Yes, if you running the EMQX in docker but running the server in the host, the URL must be the host IP, not the localhost(127.0.0.1), they are in different networks.

**Reply by vibhubithar** [reply]
yes i forgot about that for a second as i tried to make sure EMQX was correctly set, now i can see payload in my mock API server

Now trying to figure out why payloads with no values are coming through

**Reply by vibhubithar** [reply]
thanks a lot for the help :) 

this is what makes a product stand out
